File: Leetcode/easy/isIsomorphic.py
class Solution:
    def isIsomorphic(self, s: str, t: str) -> bool:
        # Time: O(N)
        # Space: O(1) - ASCII fixed at 256

        map_s_to_t = {}
        map_t_to_s = {}

        for char1, char2 in zip(s, t):
            if (char1 not in map_s_to_t) and (char2 not in map_t_to_s):
                map_s_to_t[char1] = char2
                map_t_to_s[char2] = char1
            try:
                if map_s_to_t[char1] != char2 or map_t_to_s[char2] != char1:
                    return False
            except KeyError:
                return False

        return True

'''

Runtime
35 ms
Beats
98.25%
Memory
14.3 MB
Beats
21.44%

'''

File: Leetcode/easy/isIsomorphic.java
class Solution {
    public boolean isIsomorphic(String s, String t) {
        int[] mapS_T = new int[256];
        Arrays.fill(mapS_T, -1);
        int[] mapT_S = new int[256];
        Arrays.fill(mapT_S, -1);

        for (int i = 0; i < s.length(); i++) {
            char char1 = s.charAt(i);
            char char2 = t.charAt(i);
            if (mapS_T[char1] == -1 && mapT_S[char2] == -1) {
                mapS_T[char1] = char2;
                mapT_S[char2] = char1;
            }
            else if (!(mapS_T[char1] == char2 && mapT_S[char2] == char1)) {
                return false;
            }
        }
        return true;
    }
}

/*

Runtime
3 ms
Beats
99.37%
Memory
42.5 MB
Beats
77.98%

 */
